libp2p/NodeTable.h

```diff
@@ -0,0 +1,354 @@
+		TimePoint modified;
+		std::list<std::weak_ptr<NodeEntry>> nodes;
+	};
+	
+public:
+	NodeTable(ba::io_service& _io, KeyPair _alias, uint16_t _port = 30300);
+	~NodeTable();
+	
+	/// Constants for Kademlia, mostly derived from address space.
+	
+	static unsigned const s_addressByteSize = sizeof(NodeId);				///< Size of address type in bytes.
+	static unsigned const s_bits = 8 * s_addressByteSize;					///< Denoted by n in [Kademlia].
+	static unsigned const s_bins = s_bits - 1;								///< Size of m_state (excludes root, which is us).
+	static unsigned const s_maxSteps = boost::static_log2<s_bits>::value;	///< Max iterations of discovery. (doFindNode)
+	
+	/// Chosen constants
+	
+	static unsigned const s_bucketSize = 16;		///< Denoted by k in [Kademlia]. Number of nodes stored in each bucket.
+	static unsigned const s_alpha = 3;				///< Denoted by \alpha in [Kademlia]. Number of concurrent FindNode requests.
+	
+	/// Intervals
+	
+	boost::posix_time::milliseconds const c_evictionCheckInterval = boost::posix_time::milliseconds(75);	///< Interval at which eviction timeouts are checked.
+	std::chrono::milliseconds const c_reqTimeout = std::chrono::milliseconds(300);						///< How long to wait for requests (evict, find iterations).
+	std::chrono::seconds const c_bucketRefresh = std::chrono::seconds(3600);							///< Refresh interval prevents bucket from becoming stale. [Kademlia]
+	
+	static unsigned dist(NodeId const& _a, NodeId const& _b) { u512 d = _a ^ _b; unsigned ret; for (ret = 0; d >>= 1; ++ret) {}; return ret; }
+	
+	void join();
+	
+	NodeEntry root() const { return NodeEntry(m_node, m_node.publicKey(), m_node.endpoint.udp); }
+	std::list<NodeId> nodes() const;
+	std::list<NodeEntry> state() const;
+	
+	NodeEntry operator[](NodeId _id);
+	
+	
+protected:
+	/// Repeatedly sends s_alpha concurrent requests to nodes nearest to target, for nodes nearest to target, up to s_maxSteps rounds.
+	void doFindNode(NodeId _node, unsigned _round = 0, std::shared_ptr<std::set<std::shared_ptr<NodeEntry>>> _tried = std::shared_ptr<std::set<std::shared_ptr<NodeEntry>>>());
+
+	/// Returns nodes nearest to target.
+	std::vector<std::shared_ptr<NodeEntry>> findNearest(NodeId _target);
+	
+	void ping(bi::udp::endpoint _to) const;
+	
+	void ping(NodeEntry* _n) const;
+	
+	void evict(std::shared_ptr<NodeEntry> _leastSeen, std::shared_ptr<NodeEntry> _new);
+	
+	void noteNode(Public const& _pubk, bi::udp::endpoint const& _endpoint);
+	
+	void noteNode(std::shared_ptr<NodeEntry> _n);
+	
+	void dropNode(std::shared_ptr<NodeEntry> _n);
+	
+	NodeBucket& bucket(NodeEntry const* _n);
+
+	/// General Network Events
+	
+	void onReceived(UDPSocketFace*, bi::udp::endpoint const& _from, bytesConstRef _packet);
+	
+	void onDisconnected(UDPSocketFace*) {};
+	
+	/// Tasks
+	
+	void doCheckEvictions(boost::system::error_code const& _ec);
+	
+	void doRefreshBuckets(boost::system::error_code const& _ec);
+
+#ifndef BOOST_AUTO_TEST_SUITE
+private:
+#else
+protected:
+#endif
```

Comment:
Maybe, but it's not my intention. I'd like to see this removed 1-2 PRs from now.